Implement process for ion-js cli

desaikd opened the pull request:
**Description:**

As part of ion-test-driver, this commit is for ion-js-cli process. It supports:

 - --output(-o) : rewrite given input files to specified output format
 - --output-fomrat(-f): specify ion output format
 - --error-report(-e): generate error report to specified location 


 
**Completeted Tasks:**

1. Implement processFiles function for binary/text/pretty formats to rewrite given input files.
2. Implement processEvents for generating event stream out of ion stream.
3. Supports help menu with all options and usage
4. Supports multiple input files rewritten into same output file

**TO DO:**

- [X] Support symbol[READ | WRITE | STATE] for error-report.
- [X] Support multiple input files for error-report.
- [X] Support None output format 
- [ ] Implement other options --catalog, --imports, --filter

**Test:**

tested using multiple input files for all output formats and events.
e.g. ts-node process example.ion --output-format text --output output.ion --error-report error_report

tgregg commented on src/Process.ts:
```diff
@@ -0,0 +1,133 @@
+    }
+}).argv;
+
+class ProcessArgs {
```
These args will apply to other commands too, not just process.

tgregg commented on src/Process.ts:
```diff
@@ -0,0 +1,133 @@
+}
+
+/** Error structure for error report */
+class ProcessError {
```
This ErrorReport structure applies to all commands, not just process.

tgregg commented on src/Process.ts:
```diff
@@ -0,0 +1,133 @@
+ *
+ *  Create error report through --error-report and output file specified by --output
+ */
+class Process {
```
Process is just one command available via the CLI. Compare is another. They should be accessible via the same executable. Will this support that?

desaikd replied:
No. This will not support that. Hence modularizing the code by creating another class to support both compare and process.

desaikd commented on src/Process.ts:
```diff
@@ -0,0 +1,133 @@
+class ProcessArgs {
+    inputFiles: Array<string>;
+    outputFile: any;
+    outputFormatName: string;
```
Yes. Changing it in next commit.

tgregg commented on src/Cli.ts:
```diff
@@ -0,0 +1,89 @@
+    .commandDir(__dirname,{
+        extensions: ['ts'],
+    })
+    .options({
```
How do you differentiate between the process command and other commands?

desaikd replied:
To differentiate between commands I have used a constant called command and a handler for the command file like Process.ts(8-16) which has a handler stating to call process.(Yargs support having a command directory in which you can have command files specified using a command and a handler constants).

tgregg commented on src/OutputFormat.ts:
```diff
@@ -0,0 +1,33 @@
+            case OutputFormat.EVENTS:
+                return makePrettyWriter();
+            case OutputFormat.NONE:
+                return makePrettyWriter();
```
It looks like makePrettyWriter is just a placeholder here because NONE causes nothing to be written. Is that right? If that's the case, is there a more obvious placeholder that could be used (like null)? I think a comment would be helpful, as a minimum.

tgregg commented on src/Process.ts:
```diff
@@ -0,0 +1,104 @@
+
+    processFiles(ionOutputWriter: Writer, args: IonCliCommonArgs): void {
+        for (let path of args.getInputFiles()) {
+                this.processFile(ionOutputWriter, args, path);
+            }
```
It looks like these two lines ended up with one extra level of indentation.

zslayton commented on src/Process.ts:
```diff
@@ -0,0 +1,104 @@
+import {Writer} from "./IonWriter";
+import {ErrorType, IonCliCommonArgs, IonCliError} from "./Cli";
+import {makeReader, Reader} from "./Ion";
+
```
```suggestion
// The `command`, `describe`, and `handler` exports below are part of the yargs command module API
// See: https://github.com/yargs/yargs/blob/master/docs/advanced.md#providing-a-command-module
```

zslayton commented on src/OutputFormat.ts:
```diff
@@ -0,0 +1,33 @@
+    makeTextWriter
+} from './Ion';
+
+export enum OutputFormat {
```
This could use a `typedoc` comment.

zslayton commented on src/Process.ts:
```diff
@@ -0,0 +1,104 @@
+
+/**
+ *  Read the input file(s) and re-write in the format
+ *  specified by --output-format
```
Nit: The newline in the middle of this comment is probably unnecessary.

desaikd commented on src/Cli.ts:
```diff
@@ -0,0 +1,89 @@
+}
+
+const argv = yargs
+    .commandDir(__dirname,{
+        extensions: ['ts'],
+    })
```
Yes. I think adding this would be a good idea tomato code readable.

desaikd commented on src/Cli.ts:
```diff
@@ -0,0 +1,89 @@
+        // create error output stream (DEFAULT: stderr)
+        this.errorReportFile = argv["error-report"] ? fs.createWriteStream(<string>argv["error-report"], {flags: 'w'}) : process.stderr;
+        this.outputFormatName = <OutputFormat>argv["output-format"];
+        this.inputFiles = <Array<string>>argv["inputfile"];
```
Yes. will change that in the next commit.

zslayton commented on src/IonEventStream.ts:
```diff
@@ -149,46 +157,67 @@ export class IonEventStream {
+            let currentContainerIndex: number[] = [];
+            while (true) {
+                if (this.reader.isNull()) {
+                    this.eventStream.push(this.eventFactory.makeEvent(IonEventType.SCALAR, tid!, this.reader.fieldName(), this.reader.depth(), this.reader.annotations(), true, this.reader.value()));
```
Some of these lines are very wide; could you see if your IDE has an auto-formatter? 

Note: if you autoformat the code and it changes a huge number of lines, don't worry about it. We'll finish this PR and then we can open a separate PR to do sweeping, formatting-only changes. Once that's done we can figure out how to add formatting requirements to our build pipeline.

desaikd commented on src/Cli.ts:
```diff
@@ -1,27 +1,28 @@
 export class IonCliCommonArgs {
     inputFiles: Array<string>;
-    outputFile: any;
+    outputFile: WriteStream | NodeJS.WriteStream;
```
https://nodejs.org/en/knowledge/advanced/streams/how-to-use-stream-pipe/. I am using process.stdout for printing to console if no output file specified. That is why I needed both writestreams.

zslayton replied:
Intuitively, I would expect there to be an interface for their common functionality that we could code to. The [documentation for writable streams](https://nodejs.org/api/stream.html#stream_writable_streams) seems to indicate that [`Writable`](https://nodejs.org/api/stream.html#stream_class_stream_writable) might fit that bill?

desaikd commented on src/IonEventStream.ts:
```diff
@@ -105,6 +112,7 @@ export class IonEventStream {
 
             }
         }
+        writer.close();
```
So for embedded streams it ends with stream_end for each stream embedded in that. I am not creating the writer every time but using the same for all of them.

desaikd commented:
**Commit**: e200c9c

- Support compare command 
  -  `compare`  method for basic comparison 
  - `compareEquivs` for equivs, non-equips and equiv-timeline comparison.

**Commit** : fde92fc

- fixed suggested changes
- solved merge conflicts

**Commit**: 441a1c8

- write to file only once. (remove duplicate writes) 

**TO DO:**

- Investigate on the CI build fail.

desaikd commented:
**Commit**: ff1aa82

   

- `IonEventStream` 

   - added parsing for containers and eventstreams to `compareEquivs` method. (`parseContainer` and `parseEventStream`)
   - added `equivs` which return boolean answer from `compareEquivs`.
   - modified the `compareEquivs` signature to take `comparisonType` argument and made the `comparisonReport` optional.
- `equivsTest`
  - used `equivs` inside this method in` iontests.ts`
- `IonEvent`
     - small bug fixes for comparison. (`valueCompare`)

tgregg commented on src/IonEventStream.ts:
```diff
@@ -266,7 +302,7 @@ export class IonEventStream {
                 return;
             }
-            if (tid === IonTypes.SEXP || IonTypes.LIST && this.reader.annotations()[0] === "embedded_documents") {
+            if (tid === IonTypes.SEXP && this.reader.annotations()[0] === "embedded_documents") {
```
Shouldn't this be

```
if ((tid === IonTypes.SEXP || tid === IonTypes.LIST) && this.reader.annotations()[0] === "embedded_documents") {
```

desaikd replied:
yes. Will change

desaikd commented on src/IonEventStream.ts:
```diff
@@ -189,29 +191,39 @@ export class IonEventStream {
     }
 
-    compareEquivs(expected: IonEventStream, comparisonReport: IonComparisonReport): ComparisonResult {
+    equivs(expected: IonEventStream, comparisonType: ComparisonType): boolean {
```
Sure.

zslayton commented on src/IonEventStream.ts:
```diff
@@ -138,6 +138,8 @@ export class IonEventStream {
         let actualIndex: number = 0;
         let expectedIndex: number = 0;
+        if(this.eventStream.length != expected.eventStream.length)
+            return new ComparisonResult(ComparisonResultType.ERROR, "Both event streams of different length");
```
Nit: could you wrap the body of the `if` in braces?

zslayton commented on src/IonEventStream.ts:
```diff
@@ -189,29 +191,39 @@ export class IonEventStream {
                 || !(expectedEvent.ionType == IonTypes.LIST || expectedEvent.ionType == IonTypes.SEXP)) {
                 throw new Error("Comparison sets must be lists or s-expressions.");
-            } else if(this.isEmbedded as any ^ expected.isEmbedded as any) {
+            } else if((actualEvent.annotations[0] === 'embedded_documents') as any ^ (expectedEvent.annotations[0] === 'embedded_documents') as any) {
```
> ```ts
> actualEvent.annotations[0] === 'embedded_documents') as any
> ```

Why do the boolean expressions need to be cast to `any`?

Since we check for 
```ts
someEvent.annotations[0] === 'embedded_documents'
```
several times in this code block, it might be worth making an `isEmbedded(event): boolean` helper function so the intent is obvious.

desaikd replied:
For XOR operation in TS required both types to be any or number. it doesn't work with type boolean somehow.
And yes, will add isEmbedded as a helper function.

zslayton commented on src/IonEventStream.ts:
```diff
@@ -240,23 +252,47 @@ export class IonEventStream {
                     if ((comparisonType == ComparisonType.EQUIVS || comparisonType == ComparisonType.EQUIV_TIMELINE)
                         && eventResult.result == ComparisonResultType.NOT_EQUAL) {
-                        comparisonReport.writeComparisonReport(ComparisonResultType.NOT_EQUAL, eventResult.message, i + 1, j + 1);
+                        if(comparisonReport)
```
Same thought throughout regarding `if` braces.

zslayton commented on src/IonEventStream.ts:
```diff
@@ -240,23 +252,47 @@ export class IonEventStream {
+        for (let j = 0; j < event.ionValue.length - 1; j++) {
+            container.push(event.ionValue[j]);
+            if (event.ionValue[j].eventType === IonEventType.CONTAINER_START) {
+                j += event.ionValue[j].ionValue.length
```
Could you add a semicolon to this line for consistency? (We really should add an automated style checker on this repo.)

desaikd replied:
Yes. As you mentioned earlier should we do it in a separate PR?

zslayton replied:
> As you mentioned earlier should we do it in a separate PR?

I suggest making a new PR any time you find yourself adding substantial new functionality. For example, I think this PR could have been broken up into one PR for the `yargs` framework and another PR for the `compare`/`equivs` implementation.

Incorporating feedback (style changes, refactoring, etc) can all be done as new commits in the same PR.

desaikd commented:
**Commit**: 9c1d0d2

- `IonEventStream` 

   - changed the way embedded streams were handled.
   - added doctype comment for `equivs` and `compareEquivs`.
  
- `Process`
  - changed the way embedded streams were handled to make it work properly for `List`. (`writeToEventStream`)

- Previously suggested changes

zslayton commented on src/IonEventStream.ts:
```diff
@@ -206,17 +230,19 @@ export class IonEventStream {
                 || !(expectedEvent.ionType == IonTypes.LIST || expectedEvent.ionType == IonTypes.SEXP)) {
                 throw new Error("Comparison sets must be lists or s-expressions.");
-            } else if((actualEvent.annotations[0] === 'embedded_documents') as any ^ (expectedEvent.annotations[0] === 'embedded_documents') as any) {
+            } else if(this.isEmbedded(actualEvent) as any ^ this.isEmbedded(expectedEvent) as any) {
```
Why does the boolean expression `this.isEmbedded(...)` need to be cast `as any`?

desaikd replied:
^ operator can only be applied to the number type, according to the TypeScript specification.

zslayton commented on src/IonEventStream.ts:
```diff
@@ -206,17 +230,19 @@ export class IonEventStream {
                 throw new Error("Both streams should be embedded streams.");
             }
 
+            // both containers has type any as it maybe IonEventStream[] or IonEvent[]
+            // depending on the type compare method will be called
```
Since both of these are pretty limited in scope, could we use the `IonEventStream[] | IonEvent[]` type union here? If this adds a lot of complexity, feel free to say no.

desaikd replied:
Yah. Actually I tried to do the Union type but it would add complexity to the method. So I haven't added union type here.

zslayton commented on src/Process.ts:
```diff
@@ -75,10 +77,33 @@ export class Process {
-            eventStream.writeEventStream(ionOutputWriter);
+            ionOutputWriter.writeSymbol(this.EVENT_STREAM);
+            let writeEvents: IonEvent[] = [];
+            for(let i =0; i < eventStream.getEvents().length; i++) {
```
Nit:

```suggestion
            for(let i = 0; i < eventStream.getEvents().length; i++) {
```

zslayton commented on src/Process.ts:
```diff
@@ -75,10 +77,33 @@ export class Process {
+            for(let i =0; i < eventStream.getEvents().length; i++) {
+                let event = eventStream.getEvents()[i];
+
+                //process event stream into event stream
+                if(eventStream.isEventStream) {
+                    writeEvents = eventStream.getEvents();
+                    break;
+                }
```
It looks like this case could be lifted out above the `for` loop? Consider making a helper function for collecting input stream events into an array.

```ts
if (eventStream.isEventStream) {
   writeEvents = eventStream.getEvents();
} else {
   writeEvents = collectInputStreamEvents(...); // <-- new helper function
}

for(let event of writeEvents) {
   event.write(ionOutputWriter);
}
```

That will simplify this function's responsibilities and make it easier to understand.

zslayton commented on src/Process.ts:
```diff
@@ -75,10 +77,33 @@ export class Process {
+                // processes input stream(text or binary) into event stream
+                writeEvents.push(event);
+
+                if(event.annotations[0] == 'embedded_documents') {
```
Another use case for `isEmbedded`?
